File: analysis/algorithms/representation/py/vector_magnitude.py
from __future__ import annotations

import numpy as np
from analysis.scripts.blocks import BlockResult, Packet, BlockManifest


class VectorMagnitudeBlock:
    manifest = BlockManifest(
        block_id="representation.vector_magnitude",
        group="representation",
        language="py",
        entrypoint="analysis.algorithms.representation.py.vector_magnitude:BLOCK",
        input_kinds=["raw_window"],
        output_ports={"primary": "series"},
        stateful=False,
    )

    def run(self, input_packets, params, state):
        source = input_packets["source"][0]
        s = source.data["series"]
        x, y, z = np.array(s["x"]), np.array(s["y"]), np.array(s["z"])
        mag = np.sqrt(x**2 + y**2 + z**2).tolist()
        return BlockResult(outputs={"primary": [Packet(kind="series", data={"values": mag}, axis="magnitude", sample_rate_hz=source.sample_rate_hz)]})


BLOCK = VectorMagnitudeBlock()

File: analysis/algorithms/representation/c/vector_magnitude.c
#include "../../../../analysis/scripts/c_block_protocol.h"
#include "../../../../tests/algorithms/c_stroke_rate/filters.h"

int main(void)
{
    char line[8192];
    while (analysis_protocol_read_line(line, (int)sizeof(line)))
    {
        if (!analysis_protocol_write_result("{\"outputs\":{\"primary\":[]},\"state\":{},\"diagnostics\":{}}"))
        {
            return 1;
        }
    }
    return 0;
}
